pkg/controller/metric.go

```diff
@@ -30,6 +33,39 @@ var (
 )
 
 func init() {
-	prometheus.MustRegister(errorGauge)
+	prometheus.MustRegister(errorCounterVec)
 	prometheus.MustRegister(eventHistogram)
 }
+
+func errorMetricLabels(obj runtime.Object) prometheus.Labels {
+	labels := make(map[string]string)
+
+	t, err := meta.TypeAccessor(obj)
+	if err != nil {
+		return labels
+	}
+
+	m, err := meta.Accessor(obj)
+	if err != nil {
+		return labels
+	}
+
+	labels["cr_kind"] = t.GetKind()
+	labels["cr_name"] = m.GetName()
+
+	if v, exists := m.GetLabels()["cluster.x-k8s.io/cluster-name"]; exists {
+		labels["cluster_id"] = v
+	} else if v, exists := m.GetLabels()["giantswarm.io/cluster"]; exists {
+		labels["cluster_id"] = v
+	} else {
+		labels["cluster_id"] = ""
+	}
+
+	if v, exists := m.GetLabels()["release.giantswarm.io/version"]; exists {
+		labels["release_version"] = v
+	} else {
+		labels["release_version"] = ""
+	}
+
+	return labels
```

Comment:
I have mixed feelings when it comes to those labels. They are GS specific (maybe besides cluster api ones) which is IMO wrong direction for a library. They also increase metrics cardinality greatly since cluster IDs and CR names are unbound. Also `cr_`. Not only CRs are reconciled. We sometimes are reconciling some builtin objects.

Reply:
I understand that, but on the other hand the error metric is pretty much useless without those labels and they cannot be injected outside of `operatorkit`. I'm not worried about cardinality at all since these metrics are only emitted on errors.

Reply:
How about labels: `object_kind`, `object_name` and dropping `cluster_id` and `release_version` ? Then it's not bound to us nor CR concept and in Cluster API the CR name is currently equal with Cluster ID.

Reply:
That's not the case for all the CRs, like `MachinePool`.

Reply:
Oh, that's true.

Reply:
`object_kind`, `object_name`, `object_namespace` would be fine with me in terms of being generic. In terms of cardinality I'm not convinced. If something reconciling pods is deployed with a bug then we can end up with thousands of additional metrics.

But let's focus on a problem. Why having those labels would make things useful? If it is specific problem in your team maybe writing a handler wrapper outside operatorkit that would emit metrics you like is an option?